### ISSUE TITLE
broken, misaligned interface lines while using bpytop

### ISSUE THREAD

**Issue by matkoniecz** [lxqt/qterminal#710]
Quick check whatever this version qualifies for bug reports (I am using relatively recent version, not some 4 year old LTS)
```
qterminal -v
0.14.1
```

##### Expected Behavior

bpytop works or fails due to internal bugs

##### Current Behavior

See lines on the right side
![screen02](https://user-images.githubusercontent.com/899988/95683894-40aaf600-0bee-11eb-9cf5-27053e3883fe.png)

##### Possible Solution

No idea at all.

##### Steps to Reproduce (for bugs)

1. `pip3 install bpytop --upgrade`
2. `bpytop`

##### Context

I tried using `bpytop`

I tried reporting bug in https://github.com/aristocratos/bpytop/issues/142

It was described as a terminal bug

##### System Information

* Distribution & Version: Lubuntu 20.04.1 LTS
* Kernel: how can I check this?
* Qt Version:  how can I check this?
* qtermwidget Version:  how can I check this?
* lxqt-build-tools Version: how can I check this?
* Package version: QTerminal 0.14.1
 - bpytop version: `bpytop -v` `bpytop version: 1.0.42`
 - psutil version: `bpytop -v` (version 5.7.0 or above is required) `psutil version: 5.7.2`
 - Terminal used: QTerminal 0.14.1
 - Font used: Ubuntu Mono 14 pt (default for this terminal)
 - Python version, `python3 --version` (version 3.6 or above is required): Python 3.8.5

**Comment by tsujan** [comment]
You may need to upgrade your QTerminal; 0.14.1 is outdated.

![bpytop](https://user-images.githubusercontent.com/981076/95688748-4b2eb500-0c19-11eb-82e2-a0ed8cf100b8.png)

**Comment by matkoniecz** [comment]
Thanks for testing!

Is there safe method to override packaged `qterminal`?

Standard uninstall wants to take `lxqt-core` with it what, I guess, is not a good idea.

```
The following packages will be REMOVED
  lxqt-core qterminal
```

**Comment by tsujan** [comment]
It's better to wait until your distro to upgrade its package; otherwise, you might need to do complex things and break your installation. I think the next version of Ubuntu will have 0.15.0 (although 0.16.0 will be released within a month).

I didn't close the issue; waiting for @yan12125 to test it too.

**Comment by matkoniecz** [comment]
> I think the next version of Ubuntu will have 0.15.0 (although 0.16.0 will be released within a month).

Though it will not be LTS and will push me into frequent system upgrade treadmill until early 2022 :/ (per https://ubuntu.com/about/release-cycle )

Not sure what is worse, I guess that system upgrades should be less invasive. Or just not using `bpytop`.

**Comment by yan12125** [comment]
I can confirm interface lines is fixed with `Preferences -> Appearance -> Use box drawing characters contained in the font` checked. That option is available in version 0.15.0 only [1].

> I think the next version of Ubuntu will have 0.15.0

As a record, 0.15.0 is already in Ubuntu 20.10 [2].

> Is there safe method to override packaged `qterminal`?

That is a Ubuntu question and I'm afraid people here cannot give you a good answer. Better to ask on some Ubuntu forums.

[1] https://github.com/lxqt/qterminal/commit/6771a9306d95577c8d66fabd1582842fb4673f76
[2] https://packages.ubuntu.com/source/groovy/qterminal

---

By the way, bpytop shows another issue - mismatched character width in some horizontal bars.

![image](https://user-images.githubusercontent.com/1937689/95705026-41429b80-0c85-11eb-90d5-8fd6eca79dcd.png)

That is most likely the same as https://github.com/lxqt/qterminal/issues/685.

Closing as issues are either resolved or duplicated.

**Comment by tsujan** [comment]
> I can confirm interface lines is fixed with Preferences -> Appearance -> Use box drawing characters contained in the font checked.

Strange! In my case, it is **not** checked and yet lines are correctly aligned. The temperatures are correctly shown too.

**Comment by yan12125** [comment]
Hmm, different font or font rendering methods, I guess.

**Comment by tsujan** [comment]
I'm curious; which font and font size do you use?

**Comment by yan12125** [comment]
DejaVu Sans Mono 14pt by default, but for ⣤, the symbol bpytop uses, fontconfig falls back to DejaVu Sans. I guess that non-monospace fallback font causes issues for me.
```
$ fc-match :charset=28e4                        
DejaVuSans.ttf: "DejaVu Sans" "Book"
```

**Comment by tsujan** [comment]
That's exactly the font I use, except for its size. I changed the font size from 10 to 14pt but still everything is OK with bpytop.

> I guess that non-monospace fallback font causes issues for me.

Several months ago (maybe, a year), an Arch update broke my monospace and I fixed it by adding this to `~/.config/fontconfig/fonts.conf`:

```xml
 <!-- To fix monospace, which lacked boldness after an update: -->
 <alias>
  <family>monospace</family>
  <prefer>
   <family>DejaVu Sans Mono</family>
  </prefer>
 </alias>
```

I don't remember if the problem was reported; I was happy with the fix.

**Comment by yan12125** [comment]
I have something similar
```XML
    <match>
        <test name="family">
            <string>monospace</string>
        </test>
        <edit name="family" mode="prepend">
            <string>DejaVu Sans Mono</string>
            <string>Noto Sans Mono CJK TC</string>
            <string>Noto Sans Mono CJK SC</string>
            <string>Noto Sans Mono CJK JP</string>
            <string>Noto Sans Mono CJK KR</string>
        </edit>
    </match>
```
However, if I understand it correctly (maybe not), DejaVu Sans Mono does not include U+28E4 (⣤).
```
$ fc-list --format="%{family} %{charset}\n" "DejaVu Sans Mono"
DejaVu Sans Mono 20-7e a0-1c3 1cd-1e3 1e6-1f0 1f4-1f6 1f8-1f9 1fc-221 224-241 243-245 24c-24d 250-2b9 2bb-2c1 2c6-2c9 2cc-2d3 2d6-2de 2e0-2e9 2ee 2f3 300-33f 343 358 361 374-377 37a-37f 384-38a 38c 38e-3a1 3a3-3ce 3d0-3e1 3f0-45f 462-463 472-473 490-49b 4a2-4a5 4aa-4b3 4ba-4bb 4c0-4c4 4c7-4c8 4cb-4cc 4cf-4f9 510-511 51a-51d 531-556 559-55f 561-587 589-58a 606-607 609-60a 60c 615 61b 61f 621-63a 640-655 65a 660-66d 674 679-67b 67e-680 683-684 686-687 691 698 6a4 6a9 6af 6be 6cc 6f0-6f9 e3f e81-e82 e84 e87-e88 e8a e8d e94-e97 e99-e9f ea1-ea3 ea5 ea7 eaa-eab ead-eb9 ebb-ebc ec8-ecd 10d0-10fc 1d02 1d08-1d09 1d14 1d16-1d17 1d1d-1d1f 1d2c-1d2e 1d30-1d3c 1d3e-1d5b 1d62-1d65 1d77-1d78 1d7b 1d85 1d9b-1db7 1db9-1dbf 1e00-1e13 1e18-1e2d 1e30-1e4d 1e54-1e63 1e68-1e79 1e7c-1e99 1e9b 1e9f-1ea1 1eac-1ead 1eb0-1eb1 1eb6-1eb9 1ebc-1ebd 1ec6-1ec7 1eca-1ecd 1ed8-1edd 1ee0-1ee5 1ee8-1eeb 1eee-1ef5 1ef8-1ef9 1f00-1f15 1f18-1f1d 1f20-1f45 1f48-1f4d 1f50-1f57 1f59 1f5b 1f5d 1f5f-1f7d 1f80-1fb4 1fb6-1fc4 1fc6-1fd3 1fd6-1fdb 1fdd-1fef 1ff2-1ff4 1ff6-1ffe 2000-200a 2010-2023 2026 202f-2037 2039-203a 203c-203f 2045-2049 204b 205f 2070-2071 2074-208e 2090-209c 20a0-20b5 20b8-20ba 20bd 2102 2105 210d-210f 2115-2117 2119-211a 211d 2122 2124 2126 212a-212b 212e 2148 2150-2151 2153-215f 2189 2190-2213 2215 2217-2220 2223 2227-222d 2234-223d 2241-2269 226d-228b 228d-22a5 22b2-22b5 22b8 22c2-22c6 22cd-22d1 22da-22e9 22ef 2300-2306 2308-2315 2318-2319 231c-2321 2325-2328 232b 2335-237a 237d 2380-2383 2388-238b 2395 239b-23ae 23ce-23cf 2423 2500-262f 2638-268b 2690-269c 26a0-26a1 26b0-26b1 2701-2704 2706-2709 270c-2727 2729-274b 274d 274f-2752 2756 2758-275e 2761-2775 2794 2798-27af 27b1-27bf 27c2 27c5-27c6 27dc 27e0 27e6-27eb 27f5-27f7 2987-2988 2997-2998 29eb 29fa-29fb 2a00 2a2f 2a6a-2a6b 2b05-2b0d 2b12-2b1a 2c64 2c6d-2c70 2c75-2c77 2c79-2c7a 2c7c-2c7f 2e18 2e1f 2e22-2e25 2e2e a708-a716 a71b-a71f a722-a727 a789-a78e a790-a791 a7aa a7f8-a7f9 f6c5 fb01-fb02 fb52-fb81 fb8a-fb95 fb9e-fb9f fbaa-fbad fbe8-fbe9 fbfc-fbff fe70-fe74 fe76-fefc feff fff9-fffd 1d55a
DejaVu Sans Mono 20-7e a0-1c3 1cd-1e3 1e6-1ef 1f4-1f6 1f8-1f9 1fc-21f 221 224-241 243-245 24c-24d 250-2b9 2bb-2c1 2c6-2c9 2cc-2d3 2d6-2de 2e0-2e9 2ee 2f3 300-33f 343 358 361 374-377 37a-37f 384-38a 38c 38e-3a1 3a3-3ce 3d0-3e1 3f0-45f 462-463 472-473 490-49b 4a2-4a5 4aa-4b3 4ba-4bb 4c0-4c4 4c7-4c8 4cb-4cc 4cf-4f9 510-511 51a-51d 531-556 559-55f 561-587 589-58a e3f e81-e82 e84 e87-e88 e8a e8d e94-e97 e99-e9f ea1-ea3 ea5 ea7 eaa-eab ead-eb9 ebb-ebc ec8-ecd 10d0-10fc 1d02 1d08-1d09 1d14 1d16-1d17 1d1d-1d1f 1d2c-1d2e 1d30-1d3c 1d3e-1d5b 1d62-1d65 1d77-1d78 1d7b 1d85 1d9b-1db7 1db9-1dbf 1e00-1e13 1e18-1e2d 1e30-1e4d 1e54-1e63 1e68-1e79 1e7c-1e99 1e9b 1e9f-1ea1 1eac-1ead 1eb0-1eb1 1eb6-1eb9 1ebc-1ebd 1ec6-1ec7 1eca-1ecd 1ed8-1edd 1ee0-1ee5 1ee8-1eeb 1eee-1ef5 1ef8-1ef9 1f00-1f15 1f18-1f1d 1f20-1f45 1f48-1f4d 1f50-1f57 1f59 1f5b 1f5d 1f5f-1f7d 1f80-1fb4 1fb6-1fc4 1fc6-1fd3 1fd6-1fdb 1fdd-1fef 1ff2-1ff4 1ff6-1ffe 2000-200a 2010-2023 2026 202f-2037 2039-203a 203c-203f 2045-2049 204b 205f 2070-2071 2074-208e 2090-209c 20a0-20b5 20b8-20ba 20bd 2102 2105 210d-210f 2115-2117 2119-211a 211d 2122 2124 2126 212a-212b 212e 2148 2150-2151 2153-215f 2189 2190-2213 2215 2217-2220 2223 2227-222d 2234-223d 2241-2269 226d-228b 228d-22a5 22b2-22b5 22b8 22c2-22c6 22cd-22d1 22da-22e9 22ef 2300-2306 2308-2315 2318-2319 231c-2321 2325-2328 232b 2335-237a 237d 2380-2383 2388-238b 2395 239b-23ae 23ce-23cf 2423 2500-2600 2638-263c 263f-2647 2660-2667 2669-266f 27c2 27c5-27c6 27dc 27e0 27e6-27eb 27f5-27f7 2987-2988 2997-2998 29eb 29fa-29fb 2a00 2a2f 2a6a-2a6b 2b05-2b0d 2b12-2b1a 2c64 2c6d-2c70 2c75-2c77 2c79-2c7a 2c7c-2c7f 2e18 2e1f 2e22-2e25 2e2e a708-a716 a71b-a71f a722-a727 a789-a78e a790-a791 a7aa a7f8-a7f9 f6c5 fb01-fb02 fff9-fffd 1d55a
DejaVu Sans Mono 20-7e a0-1c3 1cd-1e3 1e6-1f0 1f4-1f6 1f8-1f9 1fc-221 224-241 243-245 24c-24d 250-2b9 2bb-2c1 2c6-2c9 2cc-2d3 2d6-2de 2e0-2e9 2ee 2f3 300-33f 343 358 361 374-377 37a-37f 384-38a 38c 38e-3a1 3a3-3ce 3d0-3e1 3f0-45f 462-463 472-473 490-49b 4a2-4a5 4aa-4b3 4ba-4bb 4c0-4c4 4c7-4c8 4cb-4cc 4cf-4f9 510-511 51a-51d 531-556 559-55f 561-587 589-58a 606-607 609-60a 60c 615 61b 61f 621-63a 640-655 65a 660-66d 674 679-67b 67e-680 683-684 686-687 691 698 6a4 6a9 6af 6be 6cc 6f0-6f9 e3f e81-e82 e84 e87-e88 e8a e8d e94-e97 e99-e9f ea1-ea3 ea5 ea7 eaa-eab ead-eb9 ebb-ebc ec8-ecd 10d0-10fc 1d02 1d08-1d09 1d14 1d16-1d17 1d1d-1d1f 1d2c-1d2e 1d30-1d3c 1d3e-1d5b 1d62-1d65 1d77-1d78 1d7b 1d85 1d9b-1db7 1db9-1dbf 1e00-1e13 1e18-1e2d 1e30-1e4d 1e54-1e63 1e68-1e79 1e7c-1e99 1e9b 1e9f-1ea1 1eac-1ead 1eb0-1eb1 1eb6-1eb9 1ebc-1ebd 1ec6-1ec7 1eca-1ecd 1ed8-1edd 1ee0-1ee5 1ee8-1eeb 1eee-1ef5 1ef8-1ef9 1f00-1f15 1f18-1f1d 1f20-1f45 1f48-1f4d 1f50-1f57 1f59 1f5b 1f5d 1f5f-1f7d 1f80-1fb4 1fb6-1fc4 1fc6-1fd3 1fd6-1fdb 1fdd-1fef 1ff2-1ff4 1ff6-1ffe 2000-200a 2010-2023 2026 202f-2037 2039-203a 203c-203f 2045-2049 204b 205f 2070-2071 2074-208e 2090-209c 20a0-20b5 20b8-20ba 20bd 2102 2105 210d-210f 2115-2117 2119-211a 211d 2122 2124 2126 212a-212b 212e 2148 2150-2151 2153-215f 2189 2190-2213 2215 2217-2220 2223 2227-222d 2234-223d 2241-2269 226d-228b 228d-22a5 22b2-22b5 22b8 22c2-22c6 22cd-22d1 22da-22e9 22ef 2300-2306 2308-2315 2318-2319 231c-2321 2325-2328 232b 2335-237a 237d 2380-2383 2388-238b 2395 239b-23ae 23ce-23cf 2423 2500-262f 2638-268b 2690-269c 26a0-26a1 26b0-26b1 2701-2704 2706-2709 270c-2727 2729-274b 274d 274f-2752 2756 2758-275e 2761-2775 2794 2798-27af 27b1-27be 27c2 27c5-27c6 27dc 27e0 27e6-27eb 27f5-27f7 2987-2988 2997-2998 29eb 29fa-29fb 2a00 2a2f 2a6a-2a6b 2b05-2b0d 2b12-2b1a 2c64 2c6d-2c70 2c75-2c77 2c79-2c7a 2c7c-2c7f 2e18 2e1f 2e22-2e25 2e2e a708-a716 a71b-a71f a722-a727 a789-a78e a790-a791 a7aa a7f8-a7f9 f6c5 fb01-fb02 fb52-fb81 fb8a-fb95 fb9e-fb9f fbaa-fbad fbe8-fbe9 fbfc-fbff fe70-fe74 fe76-fefc feff fff9-fffd 1d55a 1d670-1d6a3 1d7f6-1d7ff
DejaVu Sans Mono 20-7e a0-1c3 1cd-1dd 1df-1e3 1e6-1ef 1f4-1f6 1f8-1f9 1fc-21f 221 224-241 243-245 24c-24d 250-2b9 2bb-2c1 2c6-2c9 2cc-2d3 2d6-2de 2e0-2e9 2ee 2f3 300-33f 343 358 361 374-377 37a-37f 384-38a 38c 38e-3a1 3a3-3ce 3d0-3e1 3f0-45f 462-463 472-473 490-49b 4a2-4a5 4aa-4b3 4ba-4bb 4c0-4c4 4c7-4c8 4cb-4cc 4cf-4f9 510-511 51a-51d 531-556 559-55f 561-587 589-58a e3f e81-e82 e84 e87-e88 e8a e8d e94-e97 e99-e9f ea1-ea3 ea5 ea7 eaa-eab ead-eb9 ebb-ebc ec8-ecd 10d0-10fc 1d02 1d08-1d09 1d14 1d16-1d17 1d1d-1d1f 1d2c-1d2e 1d30-1d5b 1d62-1d65 1d77-1d78 1d7b 1d85 1d9b-1db7 1db9-1dbf 1e00-1e13 1e18-1e2d 1e30-1e4d 1e54-1e63 1e68-1e79 1e7c-1e99 1e9b 1e9f-1ea1 1eac-1ead 1eb0-1eb1 1eb6-1eb9 1ebc-1ebd 1ec6-1ec7 1eca-1ecd 1ed8-1edd 1ee0-1ee5 1ee8-1eeb 1eee-1ef5 1ef8-1ef9 1f00-1f15 1f18-1f1d 1f20-1f45 1f48-1f4d 1f50-1f57 1f59 1f5b 1f5d 1f5f-1f7d 1f80-1fb4 1fb6-1fc4 1fc6-1fd3 1fd6-1fdb 1fdd-1fef 1ff2-1ff4 1ff6-1ffe 2000-200a 2010-2023 2026 202f-2037 2039-203a 203c-203f 2045-2049 204b 205f 2070-2071 2074-208e 2090-209c 20a0-20b5 20b8-20ba 20bd 2102 2105 210d-210f 2115-2117 2119-211a 211d 2122 2124 2126 212a-212b 212e 2148 2150-2151 2153-215f 2189 2190-2213 2215 2217-2220 2223 2227-222d 2234-223d 2241-2269 226d-228b 228d-22a5 22b2-22b5 22b8 22c2-22c6 22cd-22d1 22da-22e9 22ef 2300-2306 2308-2315 2318-2319 231c-2321 2325-2328 232b 2335-237a 237d 2380-2383 2388-238b 2395 239b-23ae 23ce-23cf 2423 2500-2600 2638-263c 263f-2647 2660-2667 2669-266f 27c2 27c5-27c6 27dc 27e0 27e6-27eb 27f5-27f7 2987-2988 2997-2998 29eb 29fa-29fb 2a00 2a2f 2a6a-2a6b 2b05-2b0d 2b12-2b1a 2c64 2c6d-2c70 2c75-2c77 2c79-2c7a 2c7c-2c7f 2e18 2e1f 2e22-2e25 2e2e a708-a716 a71b-a71f a722-a727 a789-a78e a790-a791 a7aa a7f8-a7f9 f6c5 fb01-fb02 fff9-fffd 1d55a
```

**Comment by tsujan** [comment]
> DejaVu Sans Mono does not include U+28E4 (⣤).

I can see `⣤` in both FeatherPad and QTerminal, which use DejaVu Sans Mono.

The old problem I mentioned above happened exactly when Noto was upgraded.

**Comment by tsujan** [comment]
It it has any relevance, this is my complete `~/.config/fontconfig/fonts.conf` (manually edited):

```xml
<?xml version='1.0'?>
<!DOCTYPE fontconfig SYSTEM 'fonts.dtd'>
<fontconfig>
 <include ignore_missing="yes">conf.d</include>
 <match target="font">
  <edit mode="assign" name="lcdfilter">
   <const>lcddefault</const>
  </edit>
 </match>
 <match target="font">
  <edit mode="assign" name="autohint">
   <bool>false</bool>
  </edit>
 </match>
 <match target="pattern">
  <edit mode="assign" name="dpi">
   <double>96</double>
  </edit>
 </match>
 <dir>~/.fonts</dir>
 <match target="font">
  <edit mode="assign" name="hinting">
   <bool>false</bool>
  </edit>
 </match>
 <match target="font">
  <edit mode="assign" name="hintstyle">
   <const>hintnone</const>
  </edit>
 </match>
 <match target="font">
  <edit mode="assign" name="rgba">
   <const>rgb</const>
  </edit>
 </match>
 <!-- For colored emojis: -->
 <match target="pattern">
  <test name="family">
   <string>monospace</string>
  </test>
  <edit mode="append" name="family">
   <string>Noto Color Emoji</string>
  </edit>
 </match>
 <!-- To fix monospace, which lacked boldness after an update: -->
 <alias>
  <family>monospace</family>
  <prefer>
   <family>DejaVu Sans Mono</family>
  </prefer>
 </alias>
 <match target="font">
  <edit mode="assign" name="antialias">
   <bool>true</bool>
  </edit>
 </match>
</fontconfig>
```

**Comment by Jogai** [comment]
It broke again with 1.4.0

**Comment by marcusbritanicus** [comment]
I can confirm that this is broken with 2.0.0 as well, no matter what the font it. I believe the problem is not QTerminal, but qtermwidget. This problem can be easily reproduced in qtermwidget's example as well.

@cc @tsujan @yan12125

**Comment by tsujan** [comment]
Everything seems OK here:

![bpytop](https://github.com/lxqt/qterminal/assets/981076/4a279298-5554-42db-8404-3cc9f2d8e41c)

**Comment by tsujan** [comment]
A bigger screenshot:

![bpytop1](https://github.com/lxqt/qterminal/assets/981076/5b91457a-0107-433e-b342-010110c4be30)

**Comment by stefonarch** [comment]
Fine here too with git master (Qt6).

**Comment by marcusbritanicus** [comment]
This is rather quite interesting... I get different results on my PC and laptop. On my PC, I use full hinting, and on laptop I use light hinting. However, if I remember correctly, we set the internal font hinting to full (here). So fontconfig should not have any affect on this.

Also to be noted, konsole and foot render bpytop perfectly irrespective of fontconfig settings. That is, they render perfectly on both laptop and PC.

**Comment by Jogai** [comment]
Fixed in 2.0.0 :smile: 

Just so you know: the author of bpytop moved on to [btop](https://github.com/aristocratos/btop) (a c++ implementation).

**Comment by marcusbritanicus** [comment]
I think the bug was in Qt and not qtermwidget/qterminal. I use a fork of qtermwidget, and for a while now, everything looks as it should.
